diff --git a/node_modules/@react-three/fiber/dist/declarations/src/three-types.d.ts b/node_modules/@react-three/fiber/dist/declarations/src/three-types.d.ts
index 7b2a48a..12120f2 100644
--- a/node_modules/@react-three/fiber/dist/declarations/src/three-types.d.ts
+++ b/node_modules/@react-three/fiber/dist/declarations/src/three-types.d.ts
@@ -383,10 +383,10 @@ export interface ThreeElements {
     fogExp2: FogExp2Props;
     shape: ShapeProps;
 }
-declare global {
-    namespace JSX {
-        interface IntrinsicElements extends ThreeElements {
-        }
-    }
-}
+// declare global {
+//     namespace JSX {
+//         interface IntrinsicElements extends ThreeElements {
+//         }
+//     }
+// }
 export {};
