// @snort/worker-relay@1.3.0/dist/interface.js
import { v4 as uuid } from "uuid";
export class WorkerRelayInterface {
    #worker;
    #commandQueue = new Map();
    // Command timeout
    timeout = 30_000;
    /**
     * Interface wrapper for worker relay
     * @param scriptPath Path to worker script or Worker script object
     */
    constructor(scriptPath) {
        if (scriptPath instanceof Worker) {
            this.#worker = scriptPath;
        }
        else {
            const sp = scriptPath ? scriptPath : new URL("@snort/worker-relay/dist/esm/worker.mjs", import.meta.url);
            this.#worker = new Worker(sp, { type: "module" });
        }
        this.#worker.onerror = e => {
            console.error(e.message, e);
        };
        this.#worker.onmessageerror = e => {
            console.error(e);
        };
        this.#worker.onmessage = e => {
            const cmd = e.data;
            if (cmd.cmd === "reply") {
                const q = this.#commandQueue.get(cmd.id);
                q?.(cmd, e.ports);
                this.#commandQueue.delete(cmd.id);
            }
        };
    }
    async init(args) {
        return await this.#workerRpc("init", args);
    }
    async event(ev) {
        return await this.#workerRpc("event", ev);
    }
    async query(req) {
        return await this.#workerRpc("req", req);
    }
    async count(req) {
        return await this.#workerRpc("count", req);
    }
    async delete(req) {
        return await this.#workerRpc("delete", req);
    }
    async summary() {
        return await this.#workerRpc("summary");
    }
    async close(id) {
        return await this.#workerRpc("close", id);
    }
    async dump() {
        return await this.#workerRpc("dumpDb");
    }
    async wipe() {
        return await this.#workerRpc("wipe");
    }
    async forYouFeed(pubkey) {
        return await this.#workerRpc("forYouFeed", pubkey);
    }
    setEventMetadata(id, meta) {
        return this.#workerRpc("setEventMetadata", [id, meta]);
    }
    async debug(v) {
        return await this.#workerRpc("debug", v);
    }
    async #workerRpc(cmd, args) {
        const id = uuid();
        const msg = {
            id,
            cmd,
            args,
        };
        return await new Promise((resolve, reject) => {
            this.#worker.postMessage(msg);
            const t = setTimeout(() => {
                this.#commandQueue.delete(id);
                reject(new Error("Timeout"));
            }, this.timeout);
            this.#commandQueue.set(id, (v, port) => {
                clearTimeout(t);
                const cmdReply = v;
                if (cmdReply.args.error) {
                    reject(cmdReply.args.error);
                    return;
                }
                resolve(cmdReply.args);
            });
        });
    }
}
