// @snort/worker-relay@1.3.0/dist/memory-relay.js
import { EventEmitter } from "eventemitter3";
import { eventMatchesFilter } from "./types";
import { debugLog } from "./debug";
/**
 * A very simple dumb fallback relay using a flat table
 */
export class InMemoryRelay extends EventEmitter {
    #events = new Map();
    #log = (msg, ...args) => debugLog("InMemoryRelay", msg, ...args);
    init() {
        this.#log("Using in-memory relay");
        return Promise.resolve();
    }
    count(req) {
        let ret = 0;
        for (const [, e] of this.#events) {
            if (eventMatchesFilter(e, req)) {
                ret++;
            }
        }
        return ret;
    }
    summary() {
        let ret = {};
        for (const [k, v] of this.#events) {
            ret[v.kind.toString()] ??= 0;
            ret[v.kind.toString()]++;
        }
        return ret;
    }
    dump() {
        const enc = new TextEncoder();
        return Promise.resolve(enc.encode(JSON.stringify(this.#events.values())));
    }
    close() {
        // nothing
    }
    wipe() {
        this.#events = new Map();
        return Promise.resolve();
    }
    event(ev) {
        if (this.#events.has(ev.id))
            return false;
        this.#events.set(ev.id, ev);
        this.emit("event", [ev]);
        return true;
    }
    eventBatch(evs) {
        const inserted = [];
        for (const ev of evs) {
            if (this.#events.has(ev.id))
                continue;
            this.#events.set(ev.id, ev);
            inserted.push(ev);
        }
        if (inserted.length > 0) {
            this.emit("event", inserted);
            return true;
        }
        return false;
    }
    sql(sql, params) {
        return [];
    }
    req(id, filter) {
        const ret = [];
        for (const [, e] of this.#events) {
            if (eventMatchesFilter(e, filter)) {
                if (filter.ids_only === true) {
                    ret.push(e.id);
                }
                else {
                    ret.push(e);
                }
            }
        }
        return ret;
    }
    delete(filter) {
        const forDelete = this.req("ids-for-delete", { ...filter, ids_only: true });
        forDelete.forEach(a => this.#events.delete(a));
        return forDelete;
    }
    setEventMetadata(_id, _meta) {
        return;
    }
}
