// nostr-tools@1.17.0/lib/cjs/test-helpers.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// test-helpers.ts
var test_helpers_exports = {};
__export(test_helpers_exports, {
  buildEvent: () => buildEvent
});
module.exports = __toCommonJS(test_helpers_exports);
function buildEvent(params) {
  return {
    id: "",
    kind: 1,
    pubkey: "",
    created_at: 0,
    content: "",
    tags: [],
    sig: "",
    ...params
  };
}
