// blossom-drive-sdk@0.4.1/lib/FileTree/methods.js
import { TreeFile } from "./TreeFile.js";
import { TreeFolder } from "./TreeFolder.js";
export function parsePath(path) {
    if (Array.isArray(path))
        return path.filter(Boolean);
    if (Array.isArray(path))
        return path.filter(Boolean);
    else
        return path.replace(/^\//, "").split("/").filter(Boolean);
}
export function formatPath(path) {
    return "/" + path.join("/");
}
export function joinPath(path, subPath) {
    return [...parsePath(path), ...parsePath(subPath)].filter(Boolean);
}
export function dirname(path) {
    return path.slice(0, -1);
}
export function basename(path) {
    return path.slice(-1)[0];
}
export function extname(path) {
    return (typeof path === "string" ? path : formatPath(path)).match(/\.[a-zA-Z0-9]{2,5}$/)?.[0];
}
export function getPath(root, path, create = false) {
    const parts = parsePath(path);
    if (parts.length === 0)
        return root;
    const next = parts.shift();
    if (!next)
        throw new Error("Bad path");
    let entry = root.get(next);
    if (!entry) {
        if (create) {
            entry = new TreeFolder(next);
            root.set(next, entry);
        }
        else
            throw new Error(`Missing folder ${next} in ${root.name}`);
    }
    if (parts.length === 0)
        return entry;
    else {
        if (entry instanceof TreeFolder)
            return getPath(entry, parts, create);
        else
            throw new Error("Expected folder");
    }
}
export function getFolder(root, path, create = false) {
    const entry = getPath(root, path, create);
    if (entry instanceof TreeFolder)
        return entry;
    else
        throw new Error("Expected folder");
}
export function getFile(root, path) {
    const entry = getPath(root, path, false);
    if (entry instanceof TreeFile)
        return entry;
    else
        throw new Error("Expected file");
}
export function setFile(root, path, metadata) {
    const parsed = parsePath(path);
    const name = basename(parsed);
    const file = new TreeFile(name, metadata);
    getFolder(root, dirname(parsed), true).set(name, file);
    return file;
}
export function remove(root, path) {
    const entry = getPath(root, path);
    if (!entry.parent)
        throw new Error("Missing parent");
    entry.parent.remove(entry.name);
}
export function move(root, src, dest) {
    const entry = getPath(root, src);
    const destFolder = getPath(root, dirname(parsePath(dest)));
    if (!(destFolder instanceof TreeFolder))
        throw new Error("Expected Folder");
    entry.parent?.remove(entry.name);
    destFolder.set(basename(parsePath(dest)), entry);
}
/** Walks a TreeFolder and calls the provided function on each branch */
export function walkTree(branch, fn) {
    const keepGoing = fn(branch);
    if (keepGoing === false)
        return;
    if (branch instanceof TreeFolder) {
        for (const child of branch)
            walkTree(child, fn);
    }
}
/** Returns all the file types found in a TreeFolder */
export function fileTypesInTree(root) {
    const types = new Set();
    walkTree(root, (branch) => {
        if (branch instanceof TreeFile)
            types.add(branch.type);
    });
    return Array.from(types).sort();
}
